### ISSUE TITLE
coc's clangd cannot find system headers when in a nix shell

### ISSUE THREAD

**Issue by Dettorer** [neoclide/coc.nvim#3549]
This is more probably a problem with my setup than a bug in coc, but I don't know how to troubleshoot this.

## Result from CocInfo

```
## versions

vim version: NVIM v0.6.0
node version: v17.3.0
coc.nvim version: 0.0.80-fe7ba28af0
coc.nvim directory: /home/dettorer/configuration/config/nvim/plugged/coc.nvim
term: alacritty
platform: linux

## Log of coc.nvim

2022-01-05T21:12:08.222 INFO (pid:602799) [services] - registered service "languageserver.ocaml-lsp"
2022-01-05T21:12:08.223 INFO (pid:602799) [services] - registered service "languageserver.nix"
2022-01-05T21:12:08.283 INFO (pid:602799) [coc-git] - Looking for git in: git
2022-01-05T21:12:08.295 INFO (pid:602799) [services] - registered service "highlight"
2022-01-05T21:12:08.297 INFO (pid:602799) [plugin] - coc.nvim 0.0.80-fe7ba28af0 initialized with node: v17.3.0 after 133ms
2022-01-05T21:12:08.393 INFO (pid:602799) [services] - registered service "clangd"
2022-01-05T21:12:08.394 INFO (pid:602799) [services] - clangd state change: stopped => starting
2022-01-05T21:12:08.397 INFO (pid:602799) [language-client-index] - Language server "clangd" started with 602830
2022-01-05T21:12:08.430 INFO (pid:602799) [services] - clangd state change: starting => running
2022-01-05T21:12:08.434 INFO (pid:602799) [services] - service clangd started
2022-01-05T21:12:09.460 INFO (pid:602799) [attach] - receive notification: highlight []
2022-01-05T21:12:17.946 INFO (pid:602799) [attach] - receive notification: showInfo []
```

## Describe the bug

When running in a nix shell that has its own clang, coc reports "file not found" for system headers such as `stdio.h`.

## Reproduce the bug

Using the suggested minimal vim config:

  ```vim
  set nocompatible
  set runtimepath^=/path/to/coc.nvim
  filetype plugin indent on
  syntax on
  set hidden
  ```

and the following C file:

```C
#include <stdio.h>

int main(void)
{
    printf("foo\n");
    return 0;
}
```

install nix and run the following commands:

```
$ nix-shell -p clang
$ nvim -u mini.vim foo.c
$ # notice the errors, see screenshot below
$ clang foo.c -o foo
$ ./foo
foo
```

## Screenshots

![image](https://user-images.githubusercontent.com/2761682/148284835-1bede26c-ab87-4d63-af1d-50dd5e6f22d2.png)

**Comment by fannheyward** [comment]
You need to setup `compile_commands.json` or `compile_flags.txt` for your project, check clangd https://clangd.llvm.org/installation#project-setup

**Comment by baronleonardo** [comment]
@Dettorer did you solve the problem ? I have exactly this one and I couldn't solve it

**Comment by Dettorer** [comment]
@baronleonardo oh yes sorry for not getting back here. That was kind of a long debug journey that ended with an easy but surprising solution.

### TLDR
also pull `pkgs.clang-tools` in your nix-shell, but do it *before* `pkgs.clang`. That is, do `nix-shell -p clang-tools clang` instead of just `nix-shell -p clang`. And in an `mkShell` definition, write `buildInputs = with pkgs; [ <other stuff> clang-tools clang <other stuff ];` instead of just `buildInputs = with pkgs; [ <other stuff> clang <other stuff> ];`. I insist that `pkgs.clang-tools` must appear in that package list *before* `pkgs.clang`.

### Long explanation
The problem has nothing to do with coc, it's entirely on nixpkgs. Nix needs to wrap a lot of binaries in its packages so they can correctly find other files in that unusual environment that is the nix store, but the problem is that the `pkgs.clang` package does not only put the `clang` program in your PATH, but also some other *unwrapped* tools, including `clangd` (or just not properly wrapped? I don't know). So the `clangd` from my nix-shell was shadowing my system's `clangd`, but it's an unwrapped version, so it cannot find basic things from the system, like system headers. The solution is then to also pull [`pkgs.clang-tools`](https://search.nixos.org/packages?channel=unstable&show=clang-tools&from=0&size=50&sort=relevance&type=packages&query=clang-tools), a package that is meant to provide binaries like `clangd` and other tools, but does so while correctly wrapping them. Although, since both `pkgs.clang-tools` and `pkgs.clang` provide a binary called `clangd`, `pkgs.clang-tools` needs to appear first in the package list of your nix-shell, so that the path for its version of `clangd` appears before the broken one from `pkgs.clang` in the resulting PATH (otherwise, calling `clangd` would still use the broken one from `pkgs.clang`).

That looks like a weird oversight from nixpkgs so I'm not sure I understand the problem fully, but I hope that makes sense and, at least, that this solution will work for you to!
